R/intervals_support_funs.R

```
@@ -0,0 +1,295 @@
+  impute_col <- if (!is.null(impute_column)) {
+    if (!impute_column %in% colnames(data)) {
+      stop("The 'data' object does not contain the specified impute column.")
+    }
+    impute_column
+  } else if ("impute" %in% colnames(data)) {
+    "impute"
+  } else {
+    data$impute <- NA_character_
+    "impute"
+  }
+  
+  # Identify the targeted intervals based on the groups
+  if (!is.null(target_groups)) {
+    target_intervals <- dplyr::inner_join(data, target_groups, by = names(target_groups))
+  } else {
+    target_intervals <- data
+  }
+  
+  # Identify the targeted intervals based on the parameters
+  target_intervals <- target_intervals %>%
+    dplyr::filter(rowSums(dplyr::across(dplyr::any_of(target_params), ~ . == TRUE)) > 0)
+  
+  # Add the imputation method to the targeted intervals
+  new_intervals_with_impute <- target_intervals %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(param_cols), ~FALSE)) %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(target_params), ~TRUE)) %>%
+    dplyr::rowwise() %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(impute_col), ~{
+      impute_methods <- unlist(strsplit(ifelse(is.na(.data[[impute_col]]), "", .data[[impute_col]]), ","))
+      if (!allow_duplication && target_impute %in% impute_methods) {
+        # If duplication is not allowed, do not add the impute method if it already exists
+        .data[[impute_col]]
+      } else {
+        # Add the impute method after the specified position
+        impute_methods <- append(impute_methods, target_impute, after)
+        paste(impute_methods[impute_methods != ""], collapse = ",")
+      }
+    })) %>%
+    dplyr::ungroup() %>%
+    dplyr::mutate(index = if (new_rows_after_original) index + 0.5 else index + max(index)) %>%
+    as.data.frame()
+  
+  # Eliminate from the old intervals the target parameters
+  old_intervals_without_impute <- target_intervals %>%
+    dplyr::mutate(dplyr::across(dplyr::any_of(target_params), ~FALSE))
```

This will not work for parameters like `time_above` where setting it to `FALSE` would be silently converted to `0` and then it would calculate the time above 0. I think that setting to `NA` should work.